Mathlib/CategoryTheory/Galois/Decomposition.lean

```diff
@@ -0,0 +1,206 @@
+
+-/
+
+/- The trivial case if `X` is connected. -/
+private lemma has_decomp_connected_components_aux_conn (X : C) [IsConnected X] :
+    ∃ (ι : Type) (f : ι → C) (g : (i : ι) → (f i) ⟶ X) (_ : IsColimit (Cofan.mk X g)),
+    (∀ i, IsConnected (f i)) ∧ Finite ι := by
+  refine ⟨Unit, fun _ ↦ X, fun _ ↦ 𝟙 X, mkCofanColimit _ (fun s ↦ s.inj ()), ?_⟩
+  exact ⟨fun _ ↦ inferInstance, inferInstance⟩
+
+/- The trivial case if `X` is initial. -/
+private lemma has_decomp_connected_components_aux_initial (X : C) (h : IsInitial X) :
+    ∃ (ι : Type) (f : ι → C) (g : (i : ι) → (f i) ⟶ X) (_ : IsColimit (Cofan.mk X g)),
+    (∀ i, IsConnected (f i)) ∧ Finite ι := by
+  refine ⟨Empty, fun _ ↦ X, fun _ ↦ 𝟙 X, ?_⟩
+  use mkCofanColimit _ (fun s ↦ IsInitial.to h s.pt) (fun s ↦ by aesop)
+    (fun s m _ ↦ IsInitial.hom_ext h m _)
+  refine ⟨by simp only [IsEmpty.forall_iff], inferInstance⟩
+
+/- Show decomposition by inducting on `Nat.card (F.obj X)`. -/
+private lemma has_decomp_connected_components_aux (F : C ⥤ FintypeCat.{w}) [FiberFunctor F]
+    (n : ℕ) : ∀ (X : C), n = Nat.card (F.obj X) → ∃ (ι : Type) (f : ι → C)
+    (g : (i : ι) → (f i) ⟶ X) (_ : IsColimit (Cofan.mk X g)),
+    (∀ i, IsConnected (f i)) ∧ Finite ι := by
+  induction' n using Nat.strongRecOn with n hi
+  intro X hn
+  by_cases h : IsConnected X
+  exact has_decomp_connected_components_aux_conn X
+  by_cases nhi : (IsInitial X → False)
+  · obtain ⟨Y, v, hni, hvmono, hvnoiso⟩ :=
+      has_non_trivial_subobject_of_not_isConnected_of_not_initial X h nhi
+    obtain ⟨Z, u, ⟨c⟩⟩ := PreGaloisCategory.monoInducesIsoOnDirectSummand v
+    let t : ColimitCocone (pair Y Z) := { cocone := BinaryCofan.mk v u, isColimit := c }
+    have hn1 : Nat.card (F.obj Y) < n := by
+      rw [hn]
+      exact ltCardFiber_of_mono_of_notIso F v hvnoiso
+    have i : X ≅ Y ⨿ Z := (colimit.isoColimitCocone t).symm
+    have hnn : Nat.card (F.obj X) = Nat.card (F.obj Y) + Nat.card (F.obj Z) := by
+      rw [cardFiber_eq_of_iso F i]
+      exact cardFiber_coprod_eq_sum F Y Z
+    have hn2 : Nat.card (F.obj Z) < n := by
+      rw [hn, hnn, lt_add_iff_pos_left]
+      exact Nat.pos_of_ne_zero (non_zero_card_fiber_of_not_initial F Y hni)
+    let ⟨ι₁, f₁, g₁, hc₁, hf₁, he₁⟩ := hi (Nat.card (F.obj Y)) hn1 Y rfl
+    let ⟨ι₂, f₂, g₂, hc₂, hf₂, he₂⟩ := hi (Nat.card (F.obj Z)) hn2 Z rfl
+    refine ⟨ι₁ ⊕ ι₂, Sum.elim f₁ f₂, Cofan.combPairHoms g₁ g₂ v u, ?_⟩
+    use Cofan.combPairIsColimit g₁ g₂ v u hc₁ hc₂ c
+    refine ⟨fun i ↦ ?_, inferInstance⟩
+    cases i; exact hf₁ _; exact hf₂ _
+  · simp at nhi
+    obtain ⟨hi⟩ := nhi
+    exact has_decomp_connected_components_aux_initial X hi
+
+/-- In a Galois category, every object is the sum of connected objects. -/
+theorem has_decomp_connected_components (X : C) :
+    ∃ (ι : Type) (f : ι → C)
+    (g : (i : ι) → f i ⟶ X)
+    (_ : IsColimit (Cofan.mk X g)),
+    (∀ i, IsConnected (f i)) ∧ Finite ι := by
+  obtain ⟨F, ⟨hf⟩⟩ := @GaloisCategory.hasFiberFunctor C _ _
```

Comment:
To avoid repetition of this sort of `obtain` in the future, could you put a definition `GaloisCategory.getFiberFunctor` (which gives the `F`) and an instance `FiberFunctor GaloisCategory.getFiberFunctor`, in the `Basic` file?